### PR TITLE
Do not cache values in object is filter is in use. 

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/58599

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -3521,8 +3521,20 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_query_is_cacheable ) {
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php]: I don't think we should use the `$id_query_is_cacheable` variable here because only part of that is relevant: All existing usage of this variable is around caching post query results (in `post-queries` group), but here we're only caching individual post data, so that has different implications. For example, the presence of `RAND(` in the orderby query does not matter for individual post caches.

Maybe we introduce a new variable `$id_posts_are_cacheabl` for the `$cacheable_field_values` logic, and then we can rely on that here.

File: src/wp-includes/class-wp-query.php [hunk 2]
```diff
@@ -3521,8 +3521,20 @@ public function get_posts() {
...
+
+			if ( $q['update_post_meta_cache'] ) {
+				update_postmeta_cache( $post_ids );
+			}
+
+			if ( $q['update_post_term_cache'] ) {
+				$post_types = array_map( 'get_post_type', $this->posts );
+				$post_types = array_unique( $post_types );
+				update_object_term_cache( $post_ids, $post_types );
 			}
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php]: Instead of duplicating all this logic from `update_post_caches()`, can we maybe find another solution where we change `update_post_caches()` (or the underlying `update_post_cache()`) instead? Maybe we can add logic there to only cache post objects if they have certain fields present?

That way we could keep this class untouched.

**Review by felixarntz — COMMENTED** [review]
I have a few minor concerns about the implementation.

File: src/wp-includes/class-wp-query.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: I don't know how that would work reliably. The query can be filtered in multiple ways. We don't know what fields were returned in the query. Is the post_content empty in the database or was it filtered to be this way. For backward compatibility reasons,I don't want to mess with `update_post_caches`. 

To ensure that the same logic runs, which is why I copied the logic over. This change also fixes a long standing bug, where post meta / terms were not primed if `cache_results` is set to false but `update_post_meta_cache` / `update_post_term_cache` were set to true.

Comment (felixarntz) [reply src/wp-includes/class-wp-query.php]: @spacedmonkey That's fair. However, the logic you have here in lines 3543-3544 differs from what is in `update_post_caches()` slightly, and because overall that logic is somewhat duplicated, I think we can find a cleaner solution to avoid the duplicate code.

Here's what I would propose:
* Introduce a new function `update_post_term_caches( WP_Post[] $posts, string|array $post_type = 'post' )`.
* Move all logic from within the `$update_term_cache` clause in `update_post_caches()` into that function, then replace it in the original function with a call to the new `update_post_term_caches( $posts, $post_type )`.
* Then, in here we can also simply call the single function `update_post_term_caches( $this->posts, $post_type )` instead of duplicating logic to figure out the post types to pass.

This way we avoid all duplicate code, and the new function `update_post_term_caches()` becomes a post-specific wrapper for the more general `update_object_term_cache()`. Plus, it fits nicely with the other existing functions `update_post_author_caches()` and `update_post_parent_caches()`.

**Comment by spacedmonkey** [comment]
@felixarntz I loved your feedback. I have actioned all of it and improved comment and tests. Ready for review. again.

File: src/wp-includes/post.php [hunk 3]
```diff
@@ -7298,10 +7298,10 @@ function clean_post_cache( $post ) {
  *
  * @since 1.5.0
  *
- * @param WP_Post[] $posts             Array of post objects (passed by reference).
- * @param string    $post_type         Optional. Post type. Default 'post'.
- * @param bool      $update_term_cache Optional. Whether to update the term cache. Default true.
- * @param bool      $update_meta_cache Optional. Whether to update the meta cache. Default true.
+ * @param WP_Post[]    $posts               Array of post objects (passed by reference).
+ * @param string|array $post_type           Optional. Post type or an array of post types. If 'any' is passed then post type is read from post objects. . Default 'post'.
```
Comment (felixarntz) [inline src/wp-includes/post.php:7302]: Good catch!

File: src/wp-includes/post.php [hunk 4]
```diff
@@ -7311,37 +7311,39 @@ function update_post_caches( &$posts, $post_type = 'post', $update_term_cache =
...
-	foreach ( $posts as $post ) {
-		$post_ids[] = $post->ID;
+	if ( $update_term_cache ) {
+		update_post_term_caches( $posts, $post_type );
 	}
 
-	if ( ! $post_type ) {
-		$post_type = 'any';
+	if ( $update_meta_cache ) {
+		$post_ids = wp_list_pluck( $posts, 'ID' );
```
Comment (felixarntz) [inline src/wp-includes/post.php]: I would prefer if we kept using the good old `foreach` loop here.

File: src/wp-includes/post.php [hunk 5]
```diff
@@ -7311,37 +7311,39 @@ function update_post_caches( &$posts, $post_type = 'post', $update_term_cache =
...
+/**
+ * Updates the cache for the given list of post objects.
+ *
+ * @since 6.4.0
+ *
+ * @param WP_Post[]     $posts     Array of post objects.
+ * @param string|array  $post_type Optional. Post type or an array of post types. If 'any' is passed then post type is read from post objects. . Default 'any'.
+ */
+function update_post_term_caches( $posts, $post_type = 'any' ) {
+	$post_ids = wp_list_pluck( $posts, 'ID' );
```
Comment (felixarntz) [inline src/wp-includes/post.php]: See above, a `foreach` loop is probably cheaper than this, even if it's just minor.

File: src/wp-includes/post.php [hunk 6]
```diff
@@ -7311,37 +7311,39 @@ function update_post_caches( &$posts, $post_type = 'post', $update_term_cache =
...
-			update_object_term_cache( $post_ids, $ptypes );
-		}
+	if ( ! $post_type ) {
+		$post_type = 'any';
 	}
 
-	if ( $update_meta_cache ) {
-		update_postmeta_cache( $post_ids );
+	if ( 'any' === $post_type ) {
+		$ptypes = wp_list_pluck( $posts, 'post_type' );
```
Comment (felixarntz) [inline src/wp-includes/post.php]: See above, I think we should keep the old `foreach` loop for this instead.

File: src/wp-includes/post.php [hunk 7]
```diff
@@ -7311,37 +7311,39 @@ function update_post_caches( &$posts, $post_type = 'post', $update_term_cache =
...
-		}
+/**
+ * Updates the cache for the given list of post objects.
+ *
+ * @since 6.4.0
+ *
+ * @param WP_Post[]     $posts     Array of post objects.
+ * @param string|array  $post_type Optional. Post type or an array of post types. If 'any' is passed then post type is read from post objects. . Default 'any'.
+ */
+function update_post_term_caches( $posts, $post_type = 'any' ) {
```
Comment (felixarntz) [inline src/wp-includes/post.php:7335]: It would be great if you could add test coverage directly for this function, e.g. 3 tests, one for each alternative way to provide the `$post_type` parameter.

File: src/wp-includes/class-wp-query.php [hunk 8]
```diff
@@ -3528,8 +3529,17 @@ public function get_posts() {
...
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
+				update_post_cache( $this->posts );
+			}
+
+			if ( $q['update_post_meta_cache'] ) {
+				$post_ids = wp_list_pluck( $this->posts, 'ID' );
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php]: See my other comments. It's a bit cheaper to use a simple `foreach`. It's 2-3 more lines of code, but that's okay. :)

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Overall, this PR looks good now to me.

I only have one _very_ minor performance nit-pick: We're now using `wp_list_pluck()` in a few places, more times than before, and before that code was just using a `foreach`. `wp_list_pluck()`, while not expensive, instantiates an entire class and provides flexibility that we don't need for the use-cases here. So I think it would be a bit more efficient to replace all of them with a `foreach`.

Other than that, it would be great to add tests for the new function specifically.

File: src/wp-includes/class-wp-query.php [hunk 9]
```diff
@@ -3528,8 +3529,17 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
+				update_post_cache( $this->posts );
+			}
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-query.php:3534]: In the example provided on the ticket the issue was that `$this->posts` may not contain a full post object.

I think it's best to cache the entire object reqardless using something like:

```php
if ( $q['cache_results'] ) {
        if ( $id_query_is_cacheable ) {
                update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
        } else {
                $post_ids = wp_list_pluck( $this->posts, 'ID' );
                _prime_post_caches( $post_ids, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
        }
}
```

`cache_results` has always been about the main query rather than priming the posts -- the current change will end up priming the posts as each `get_post()` is called so will take it from 1 query to potentially 10.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php:3534]: Imagine this, I modify the query to return to just the ids and the excerpts. I do this because I don't need the post content. The return values get cast to posts objects here. 

```
$this->posts = array_map( 'get_post', $this->posts );
```

In my change, this post object missing fields will not be stored in cache. And the overhead of post content is not sent from the database. In your change, `_prime_post_caches` would result in another database query and the whole post object would be queries. 

When looping around the posts, globals are setup for the object stored in the posts property. 

TLDR, calling `_prime_post_caches` may result in query for post objects, which is what using the filter was trying to avoid.

File: src/wp-includes/post.php (hunk 4 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php]: It is the performance reason for this? Or just a code style one?

Comment (felixarntz) [reply src/wp-includes/post.php]: I explained in the overarching PR review comment https://github.com/WordPress/wordpress-develop/pull/5134#pullrequestreview-1617974444. A `foreach` loop is more efficient and performant than instantiating a `WP_List_Util` object with all its unnecessary flexibility.

File: src/wp-includes/post.php (hunk 7 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php:7335]: Added unit tests.

File: src/wp-includes/class-wp-query.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-query.php:3534]: I think this is a semantics discussion: We have two options here:
* Either we interpret the `cache_results` flag as "cache the query results", in which case the current way this PR is implemented makes sense. We will only cache the individual posts if their data is part of the query results, effectively getting this post caching "for free".
* Or we interpret the `cache_results` flag to also encompass "cache the query results and post results", in which case something like what you outlined @peterwilsoncc makes sense. The drawback would be that WP would run another DB query, so at that point I would question why it's even possible to query fewer post fields, if in the end all post fields are queried anyway.

I'm personally leaning towards the first option.

Comment (peterwilsoncc) [reply src/wp-includes/class-wp-query.php:3534]: @spacedmonkey @felixarntz Priming the entire object cache fixes the regression introduced in 5a497225b286964ff9302a21272731172840ed05 for WP 6.2.

Prior to that commit, the posts were primed using the IDs and an additional database query, the commit changed it so the posts were primed using the selected post data. What we (it looks the like three of us worked on it) missed in that change was when the data was filtered in the way reported in this ticket.

For a long time the individual post caches were primed regardless of the `cache_results` setting so I think it's best to continue doing that.

> ...at that point I would question why it's even possible to query fewer post fields, if in the end all post fields are queried anyway.

I don't disagree but as it's being the case for so long I'm hesitant to change it. I worry that instead of six database queries we'd end up with sites making 15 database queries because they don't realise the posts are filtered.

Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php:3534]: @peterwilsoncc How about a new parameter that allow developers to opt-in/out of storing post objects. That way we do not have to guess if this is the desired behaviour or not. 

I agree with @felixarntz, if are trying avoid loading, say large "post_content" field, then loading them anyway using `_prime_post_caches` will have negative effects. 

It is unclear to all of us, then it will be unclear to developers as well. 

Reminder, we don't need to prime post caches for the loop. The `$post` global comes from array stored in property. 

https://github.com/WordPress/wordpress-develop/blob/60d2d8ac75c4350a45e4c6970c7d0295b757f1f0/src/wp-includes/class-wp-query.php#L3669-L3670

https://github.com/WordPress/wordpress-develop/blob/60d2d8ac75c4350a45e4c6970c7d0295b757f1f0/src/wp-includes/class-wp-query.php#L3621-L3628

Not priming these most likely would not result in 15 database queries.

File: src/wp-includes/class-wp-query.php [hunk 10]
```diff
@@ -3166,7 +3166,8 @@ public function get_posts() {
 			"{$wpdb->posts}.ID",
 		);
 
-		if ( ! in_array( $fields, $cacheable_field_values, true ) ) {
+		$id_posts_are_cacheable = in_array( $fields, $cacheable_field_values, true );
+		if ( ! $id_posts_are_cacheable ) {
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php:3170]: Technically speaking, this is a potential issue currently: The posts really are only cacheable if _all_ post fields are queried. If the query queries only `ID`, or only `ID` and `post_parent`, we must not cache those results as it would be incomplete post objects.

Right now this wouldn't cause real problems since the `update_post_cache()` call below won't be reached in the other two cases anyway since for both of them (`ids` and `id=>parent`) early returns are present. But it's a risk so I'd prefer to set this variable only to true if really _all_ post fields are queried.

This is a completely separate check though, so I think we can revert this and apply the other check only where it's needed further below.
```suggestion
		if ( ! in_array( $fields, $cacheable_field_values, true ) ) {
```

File: src/wp-includes/class-wp-query.php [hunk 11]
```diff
@@ -3528,8 +3529,20 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
```
Comment (felixarntz) [inline src/wp-includes/class-wp-query.php:3532]: See my other comment, plus there's one additional thing to consider here: In order to ensure that `$this->posts` contains full post objects, `"{$wpdb->posts}.*" === $fields` must be true, _and_ we have to ensure that none of the previous filters potentially modified the post results. Our source of truth is `$this->posts` in the way that it is set in https://github.com/WordPress/wordpress-develop/blob/3bacfc9f978876549ee0e2b5068be02e006d956c/src/wp-includes/class-wp-query.php#L3337. After that, it may be modified again by certain filters etc., and we cannot assume that the post fields are still intact if any such modifications happened. So I think we should "store" the value from https://github.com/WordPress/wordpress-develop/blob/3bacfc9f978876549ee0e2b5068be02e006d956c/src/wp-includes/class-wp-query.php#L3337 in a variable `$unfiltered_posts` to compare against here. Only if the array still matches, no modifications were made and we can reliably use `$this->posts` to cache the objects.
```suggestion
			if ( $q['cache_results'] && "{$wpdb->posts}.*" === $fields && $unfiltered_posts === $this->posts ) {
```

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey I found two potential problems with the current logic to decide whether to cache the objects from `$this->posts`.

File: src/wp-includes/class-wp-query.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/class-wp-query.php:3534]: @spacedmonkey After reviewing the logic again more closely, I found two related problems which may cause `$this->posts` to not contain full post objects, see https://github.com/WordPress/wordpress-develop/pull/5134#pullrequestreview-1633962867.

I think we will need to cater for that.

Separately, I agree now with @peterwilsoncc that for other cases we should continue to have `_prime_post_caches()` in an `else` for the case the post objects are incomplete, and here's why:
* The main reason to avoid the `_prime_post_caches()` call originally was that when `$this->posts` has the complete post objects, we can avoid a database query. This is still the case here.
* Calling `_prime_post_caches()` in an `else` will only be reached in edge-cases (e.g. where post fields or the actual results are filtered), but for those I agree that the post caches should continue to be primed, since otherwise it can lead to accidental excess database queries.

Adding back the `_prime_post_caches()` call in that scenario will not realistically have a notable negative performance impact, while ensuring that such edge-cases where query fields or query results are filtered still behave the same way as they did before. The original performance goal for why the main `_prime_post_caches()` call was removed is still maintained.

**Comment by spacedmonkey** [comment]
After some great feedback from @felixarntz and @peterwilsoncc. I agree this PR has problems. I think a new approach is needed. I am going to close this PR for now, in favour of https://github.com/WordPress/wordpress-develop/pull/5251

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/e486ac07458d36182ee75091eac06f9a99a5ae2a
